auslib/db.py

```diff
@@ -2490,6 +2491,53 @@ def _putWatchdogValue(self, changed_by, value, where=None, transaction=None, dry
             super(Dockerflow, self).update(where=where, what=value, changed_by=changed_by, transaction=transaction, dryrun=dryrun)
 
 
+class EmergencyShutoff(AUSTable):
+    def __init__(self, db, metadata, dialect):
+        self.table = Table('emergency_shutoff', metadata,
+                           Column('shutoff_id', Integer, primary_key=True, autoincrement=True),
+                           Column('product', String(15), nullable=False),
+                           Column('channel', String(75), nullable=False),
+                           Column('updates_disabled', CompatibleBooleanColumn, default=False, nullable=False),
+                           Column('additional_notification_list', String(500)))
+        AUSTable.__init__(self, db, dialect, scheduled_changes=True, scheduled_changes_kwargs={"conditions": ["time"]})
+
+    def getEmergencyShutoff(self, shutoff_id):
+        emergency_shutoff = self.select(where=[self.shutoff_id == shutoff_id])
+        return emergency_shutoff[0] if emergency_shutoff else None
+
+    def getEmergencyShutoffs(self, where=None):
+        return self.select(where=where)
+
+    def insert(self, changed_by, transaction=None, dryrun=False, signoffs=None, **columns):
+        ret = super(EmergencyShutoff, self).insert(changed_by=changed_by, transaction=transaction, dryrun=dryrun, **columns)
+        return ret.inserted_primary_key[0]
+
+    def update(self, where, what, changed_by, old_data_version, transaction=None, dryrun=False, signoffs=None):
```

Comment:
Out of curiosity, did you consider removing rows altogether instead of updating them? Either is OK, I'm just wondering if you considered both.

Reply:
I do not understand your question, I think.
The update method changes enable/disable state, more than that, update does not require signoffs for disabling updates.

Reply:
From my perspective, there's two ways that shutoffs can be handled:
1) If no row exists in this table, or updates_disabled is False - the product+channel is NOT shutoff. If the row exists in the table, and updates_disabled is True - the produt+channel is shutoff. This means that the first time we shut updates off, a new row is created, and whenever we make a change after that, we update the updates_disabled column.
2) If no row exists in the table, updates are on. If a row does exist in the table, updates are off. This means that when we shut updates off, we insert a row, and when we turn them back on, we remove it.

As far as I can tell, this PR does the first, and I was wondering if you thought about the second.

Reply:
Ah, actually not! Because I thought about workflow in another way:
1 - User wants to give to product/channel the possibility to disable updates, so he creates a row with 'updates_disabled' False.
1.1 - If updates needs be disabled, just call to change 'updates_disabled' to True and Schedule an ASAP change (that requires signoffs) to go back 'updates_disabled' to False.

So in that way an admin can define what can be shut off and the others user can see what is available to be shut off.

I like that second way, is more simple. Do you think we can change to add/remove approach?

Reply:
Ah, I see what you mean - it gives a bit more control over the types of shutoffs that can be done. I think it's a bit more than is needed though - if we need to be more precise about the shutoffs that can be done we can add support for "product" and/or "channel" as an option on whichever permission we create for making emergency shutoffs.